### PR TITLE
ref: Moved `safe_forward` to  `utils.py`

### LINKED ISSUE #124: Why is `safe_forward` in `growing_container.py` ?
The function `safe_forward` is for now in `growing_container.py` but is only used in `GrowingDAG` in `growing_dag.py`. It does not seems a logical place. If there is no particular reason to have it in `growing_container.py`, I would suggest to move it to `utils.py` (or `tools.py`).

### PR DESCRIPTION
**Pull request by TheoRudkiewicz**
This location is more logical for  `safe_forward`.
This fix #124.

In addition i changed the error message in the assert changing `Found: {input.shape[1]}` by `Found: {input.shape[-1]}`. This seems more logical @stelladk can you confirm this is ok for you ?

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -346,6 +347,78 @@ def test_f1_score_functions(self) -> None:
...
+                # Test with zero input features (edge case)
+                zero_in_features = 0
+                zero_out_features = 3
+                zero_linear = nn.Linear(
+                    zero_in_features, zero_out_features, device=device
+                )
+                zero_input = torch.empty(batch_size, zero_in_features, device=device)
+
+                zero_output = safe_forward(zero_linear, zero_input)
+                self.assertShapeEqual(zero_output, (batch_size, zero_out_features))
```
Comment (stelladk) [inline tests/test_utils.py:386]:
```suggestion
                zero_in_features = 0
                zero_linear = nn.Linear(
                    zero_in_features, out_features, device=device
                )
                zero_input = torch.empty(batch_size, zero_in_features, device=device)

                zero_output = safe_forward(zero_linear, zero_input)
                self.assertShapeEqual(zero_output, (batch_size, out_features))
```

**Review by stelladk — APPROVED** [review]
